### PR TITLE
integrate(storage/agent): land reviewed mutation and RS-heal stack

### PR DESCRIPTION
**Pull request by Jurij89**
## Summary

- Restore the already-reviewed #2187/#2194/#2196/#2197 stack to the actual `integration/2052-system-record-sync` line.
- Close leased raw SPARQL mutation routes, preserve the reviewed managed-store architecture gate, and split structured mutations into focused storage-owned modules.
- Extract RS-heal materialization and scheduler policy from SWM host orchestration while preserving bounded preflight, stale checks, cancellation, and write ordering.
- This recovery PR is necessary because the stacked PRs were squash-merged into feature branches after their parents had already been squash-merged; GitHub marked them merged, but their commits did not propagate to integration.

## Related

- Integrates #2187
- Integrates #2194
- Integrates #2196
- Integrates #2197
- Builds on #2185 and #2195, which are already present in integration
- Part of #2052
- Plans: `agent-docs/plans/2026-08-04-issue-2052-system-graph-replay-remainder.md` and `agent-docs/plans/2026-08-09-system-record-followup-refactors.md`

### Managed mutation ownership

Before:

```mermaid
sequenceDiagram
    participant Feature
    participant Store as Managed store
    participant Backend
    Feature->>Store: Raw update/query mutation
    Store->>Backend: Caller-authored SPARQL
    Backend-->>Store: Result
    Store-->>Feature: Result
```

After:

```mermaid
sequenceDiagram
    participant Feature
    participant Module as Structured mutation module
    participant Store as Managed store
    participant Backend
    Feature->>Module: Typed bounded mutation
    Module->>Store: Validated structured operation
    Store->>Backend: Storage-owned mutation
    Backend-->>Store: Exact commit effect
    Store-->>Feature: Typed outcome
```

### RS-heal materialization

Before:

```mermaid
sequenceDiagram
    participant Host as SWM host
    participant Store
    Host->>Host: Build reset/copy/metadata/stamp mutations
    Host->>Store: Execute mutation sequence
    Store-->>Host: Results
```

After:

```mermaid
sequenceDiagram
    participant Host as SWM host
    participant Materializer
    participant Store
    Host->>Materializer: Graphs, UAL, roots, version, predicates
    Materializer->>Materializer: Preflight capability and chunk bounds
    Materializer->>Store: Reset marker
    loop Each bounded chunk while current
        Materializer->>Store: Copy exact subject projection
    end
    Materializer->>Store: Copy metadata and stamp completion
    Store-->>Host: Typed materialization outcome
```

## Files changed

| File | What |
|------|------|
| `packages/core/src/sparql-operation.ts` | Classify read, mutation, and unknown SPARQL operations for fail-closed admission. |
| `packages/storage/src/structured-mutation/*` | Split storage-owned mutation builders into focused modules. |
| `packages/storage/src/bounded-structured-mutation.ts` | Retain the bounded public facade over the focused modules. |
| `packages/storage/src/adapters/*` | Enforce managed ownership and structured mutation dispatch across adapters. |
| `packages/agent/src/rs-heal-materialization.ts` | Own RS-heal capability proof, preflight, stale boundaries, and ordered execution. |
| `packages/agent/src/rs-heal-store-options.ts` | Centralize RS-heal scheduler and tracing options. |
| `packages/agent/src/dkg-agent-swm-host.ts` | Delegate domain inputs while retaining discovery and orchestration. |
| `scripts/check-managed-store-raw-channels.mjs` | Reject new first-party raw managed-store mutation call sites. |
| `.github/workflows/system-record-managed-ownership.yml` | Run both the current type-contract gate and raw-channel architecture gate. |
| `packages/**/test/*.test.ts` | Preserve mutation, query-admission, wrapper, and RS-heal behavioral evidence. |

## Test plan

- [x] Run dependency-aware Agent build, including type and package-root checks.
- [x] Run `pnpm check:managed-store-raw-channels` (287 reviewed dynamic query sites, zero raw/static mutation violations).
- [x] Run Storage type-contract compilation.
- [x] Run focused structured-mutation, reserved-graph, and managed-admission suites (111/111).
- [x] Run all RS-heal materialization/decorated/host/HTTP suites (38/38).
- [x] Run Query engine admission suite (128/128).
- [x] Run Agent structured replacement wrapper suite (4/4).
- [x] Run `git diff --check` and confirm a clean worktree after restoring generated EVM deployment output.
- [ ] Confirm integration CI passes on the reconstructed stack.

### CODE REVIEW

**Comment by Jurij89** [comment]
@otReviewAgent Please perform a full merge-readiness review of the reconstructed integration stack at 68dc847e48321334588d52e9d8d9d8f1dc320ae8. Check the integration-specific conflict resolution, cumulative dependency semantics, managed-store ownership/admission behavior, structured mutation module split, RS-heal materialization boundaries, resource pressure, cancellation, and regression coverage.

File: packages/storage/src/adapters/sparql-http.ts [hunk 1]
```diff
@@ -1258,27 +1298,34 @@ export class SparqlHttpStore implements TripleStore {
...
-    this.writeGen.recordGraphWrites(touchedGraphs);
+    this.writeGen.recordGraphWrites(effects.touchedGraphs);
   }
 
   async query(sparql: string, options?: SparqlHttpQueryOptions): Promise<QueryResult> {
+    const analysis = analyzeSparqlOperation(sparql);
+    const operation = analysis.operation;
+    this.assertRawSparqlChannelAvailable({
+      kind: 'query',
+      readForm: recognizedReadOnlySparqlForm(analysis),
```
Comment (otReviewAgent) [inline packages/storage/src/adapters/sparql-http.ts:1326]: 🔴 Bug: Leased query admission can miss update text after an escaped hash in a prefixed name

**What's wrong**
The new leased-store guard trusts `recognizedReadOnlySparqlForm(analysis)` to decide whether raw `query()` may go to the endpoint. That analysis strips every `#` to end-of-line as a comment before looking for mutating keywords, but SPARQL permits `\#` as escaped punctuation inside a prefixed local name. A caller can place a legal escaped hash before trailing update text on the same line; the update keyword is removed from the analyzed string, so the guard treats the program as a read and does not enforce the promised pre-I/O refusal.

**Example**
`PREFIX ex: <urn:x:> SELECT ?s WHERE { ?s ex:foo\\# ?o } ; DROP ALL` contains an escaped `#` in a prefixed local name, so the `#` is not a SPARQL comment. The current stripper blanks from that hash through `DROP ALL`, `recognizedReadOnlySparqlForm(analysis)` returns `SELECT`, and a leased `query()` can reach `fetch` instead of throwing `MANAGED_OXIGRAPH_MUTATION_UNAVAILABLE` before I/O.

**Suggested direction**
Make comment stripping/tokenization aware of escaped `#` in SPARQL prefixed local names, or switch the leased raw-channel admission check to a tokenizer/parser that can distinguish comments from PN_LOCAL_ESC before deciding a query is a recognized read.

**Confidence note**
This relies on SPARQL's escaped local-name hash being parsed as part of a prefixed name, not as a comment. Even if the backend ultimately rejects the combined text, the new managed raw-channel contract is to refuse mutating/unknown query programs before I/O.

**For Agents**
Look in `packages/core/src/sparql-operation.ts` and the leased query tests in `packages/storage/test/reserved-internal-graph-mutation-guard.test.ts`. Preserve allowed update words in variables/prefixed names, but make escaped `\#` inside PN_LOCAL not start a stripped comment, so following top-level update text remains visible. Add a leased-store regression that spies on fetch and proves the escaped-hash read-then-update string is refused before I/O.

File: scripts/check-managed-store-raw-channels.mjs [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+} from '../packages/core/src/sparql-operation.ts';
+import ts from 'typescript';
+
+const REPOSITORY_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
+const STORAGE_PACKAGE_NAME = '@origintrail-official/dkg-storage';
+const REVIEWED_NON_STORE_MARKER = 'dkg-raw-channel-non-store';
+// Reviewed dynamic read call sites are sealed per package. A new package, call,
+// or changed query expression fails the gate and prints the exact inventory
+// that needs review before this baseline may be deliberately advanced.
+const DYNAMIC_QUERY_BASELINE = Object.freeze({
```
Comment (otReviewAgent) [inline scripts/check-managed-store-raw-channels.mjs:20]: 🟡 Issue: Replace the dynamic query hash baseline with an explicit read boundary

**What's wrong**
This script adds a 477-line bespoke TypeScript analyzer whose main escape hatch is a package-level hash of all dynamic `TripleStore.query()` expressions. That is a magic baseline, not a clean abstraction: it couples CI to incidental expression text, requires reviewers to audit a dumped inventory whenever unrelated code moves, and does not make dynamic read intent visible at the call site. The runtime ownership guard is already the authoritative safety boundary; this additional gate should make first-party architecture clearer, not introduce a second opaque maintenance surface.

**Example**
A harmless refactor such as renaming a local query variable or moving a call site can change the fingerprint and fail CI, while a real semantic change hidden behind the same dynamic argument text can keep the hash stable. That makes the gate noisy in the cases reviewers should ignore and opaque in the cases reviewers need to understand.

**Suggested direction**
Keep the architectural rule, but model it directly instead of snapshotting hundreds of dynamic calls. A small typed helper or explicit allowlist with named call sites would be easier to review, diff, and maintain than counts plus package-level hashes.

**For Agents**
Rework `scripts/check-managed-store-raw-channels.mjs` and the workflow step. Preserve the runtime leased-store refusal behavior and the prohibition on first-party raw writes. Replace the hash baseline with a simpler explicit boundary: ban typed `TripleStore.update()`, reject statically mutating `query()` calls, and route dynamic read programs through a named read-only helper or a small reviewable manifest of intentional dynamic read sites. Add/adjust a focused test or self-test that proves the scanner fails on a new raw update call and accepts intentional dynamic reads without hash churn.

File: packages/storage/src/adapters/sparql-http.ts (hunk 1 above)
Comment (Jurij89) [reply packages/storage/src/adapters/sparql-http.ts:1326]: Addressed in d504fc6d8. Comment stripping now recognizes PN_LOCAL_ESC at the hash character, so an escaped \\# remains part of the prefixed local name and trailing top-level update text remains visible to admission. The leased-store regression uses the reported SELECT-then-DROP shape, asserts MANAGED_OXIGRAPH_MUTATION_UNAVAILABLE, and proves fetch is never called. Core rebuild and the full reserved mutation guard suite pass 79/79.

File: scripts/check-managed-store-raw-channels.mjs (hunk 2 above)
Comment (Jurij89) [reply scripts/check-managed-store-raw-channels.mjs:20]: Valid architectural follow-up, already implemented and reviewed in the immediately subsequent PR #2198, which fixes #2192 and #2193. That PR replaces the package hash baseline with a checked-in per-call-site JSON inventory carrying path, symbol, expression, occurrence, and reason, plus scanner self-tests and an explicit regeneration path. This recovery PR intentionally reconstructs the prerequisite stack; duplicating #2198 here would collapse the reviewed dependency boundary. #2198 is next in the audited integration order after this PR.

**Comment by Jurij89** [comment]
@otReviewAgent Please re-review current head d504fc6d80d642780a52eaa4717dd6a69d7bbcb3 after the escaped PN_LOCAL hash admission fix. Confirm the red pre-I/O query bypass is closed, the deferred scanner-maintainability item is fully tracked by immediately subsequent PR #2198, the recovery composition remains correct, and current CI is merge-ready.

File: packages/storage/src/bounded-structured-mutation.ts [hunk 3]
```diff
@@ -680,3 +146,17 @@ export function structuredMutationTouchedGraphs(mutation: StructuredMutation): r
...
   return mutation.kind !== 'delete-subjects' || mutation.input.subjects.length > 0;
 }
+
+/** Immutable graph-scoped effects captured before a structured mutation is dispatched. */
+export interface StructuredMutationEffects {
+  readonly touchedGraphs: readonly string[];
+}
+
+/** Capture canonical effects without executing or probing a store capability. */
+export function captureStructuredMutationEffects(
```
Comment (otReviewAgent) [inline packages/storage/src/bounded-structured-mutation.ts:156]: 🟡 Issue: Bundle structured-mutation normalization, effects, and update planning

**What's wrong**
`captureStructuredMutationEffects` adds another partial abstraction rather than deleting the repeated orchestration. Callers must still know when normalization has happened, whether effects are frozen, how no-ops are represented, and how that relates to update generation. That order-sensitive contract is not encoded in the type and will get worse as new structured mutation kinds are added.

**Example**
`SparqlHttpStore.structuredMutation()` now has to remember the exact sequence: normalize at line 1291, capture effects at line 1292, build at line 1300, then check both `update` and `effects` at line 1301. Other adapters repeat the same choreography, but the agent wrapper captures from the caller object before normalization to avoid later mutation.

**Suggested direction**
Introduce one preparation API, for example `prepareStructuredMutation(mutation)`, that synchronously clones/normalizes, computes guarded/touched graphs, and returns either a no-op plan or an executable plan containing the update and effects. Then adapters and decorators consume that plan instead of independently reassembling the same sequence.

**For Agents**
Work in `packages/storage/src/bounded-structured-mutation.ts` and the `structuredMutation` implementations. Preserve current no-op behavior and pre-dispatch graph capture. Add or keep tests proving empty `delete-subjects` does not invalidate and post-call mutation of the caller object cannot redirect effects.

File: packages/core/src/sparql-operation.ts [hunk 4]
```diff
@@ -152,16 +152,116 @@ function classifySparqlOperationForm(form: SparqlDetectedOperation): SparqlOpera
...
+    mutatingKeyword: findMutatingKeyword(stripped),
   };
 }
 
+/**
+ * Canonical read-only admission policy for already-analyzed SPARQL.
+ * A recognized read form is safe only when no standalone executable update
+ * keyword remains elsewhere in the program (for example after a semicolon).
+ */
+export function recognizedReadOnlySparqlForm(
```
Comment (otReviewAgent) [inline packages/core/src/sparql-operation.ts:257]: 🟡 Issue: Make the read-only SPARQL policy genuinely canonical

**What's wrong**
The new helper centralizes the leased-store and scanner decision, but the public query validator still repeats the same read-only checks. That leaves multiple policy surfaces for a subtle SPARQL lexical rule set, which is exactly the kind of boundary that should have one owner.

**Example**
If the core helper later allows a new read form or changes mutating-keyword treatment, leased-store admission and query-engine validation can diverge unless both implementations are updated in lockstep.

**Suggested direction**
Collapse the duplicated policy into one exported admission function, or have `validateReadOnlySparql()` derive its decision from `recognizedReadOnlySparqlForm()` and the same `SparqlOperationAnalysis`. Keep formatting of user-facing reasons as a thin layer, not a second policy implementation.

**Confidence note**
This is based on the changed helper plus surrounding query guard context; it needs the author to confirm whether `validateReadOnlySparql` is intended to remain a separate policy surface.

**For Agents**
Update `packages/query/src/sparql-guard.ts` to consume the new canonical helper, or replace both with one core admission function that returns the read form plus a reason. Preserve current validation messages and tests, including update-word names and read-then-update programs.

File: packages/storage/src/bounded-structured-mutation.ts (hunk 3 above)
Comment (Jurij89) [reply packages/storage/src/bounded-structured-mutation.ts:156]: Valid maintainability follow-up, deliberately kept out of this integration recovery patch. Filed #2208 with the exact preparation-plan direction and acceptance tests: one discriminated snapshot/normalize/effects/update owner, no-op preservation, caller-mutation isolation, exact invalidation, adapter parity, and no added pressure. #2207 preserves the already-reviewed behavior while #2208 owns the no-behavior-change consolidation.

File: packages/core/src/sparql-operation.ts (hunk 4 above)
Comment (Jurij89) [reply packages/core/src/sparql-operation.ts:257]: Valid and assigned to the immediately subsequent #2198/#2193 scanner-policy PR rather than duplicated in recovery. Before #2198 enters integration, validateReadOnlySparql will derive admission from the same Core analysis/recognized-read helper while retaining its user-facing reason mapping and tests. Issue #2193 will remain open until that consolidation is present on the integration path.
